pkg/provider/kv/consul/consul.go

```diff
@@ -4,30 +4,78 @@ import (
 	"errors"
 
 	"github.com/kvtools/valkeyrie/store"
+	"github.com/traefik/traefik/v2/pkg/log"
 	"github.com/traefik/traefik/v2/pkg/provider"
 	"github.com/traefik/traefik/v2/pkg/provider/kv"
 )
 
+// providerName is the Consul provider name.
+const providerName = "consul"
+
 var _ provider.Provider = (*Provider)(nil)
 
-// Provider holds configurations of the provider.
-type Provider struct {
+// ProviderBuilder is responsible to construct namespaced instances of the Consul provider.
+type ProviderBuilder struct {
 	kv.Provider `export:"true"`
+
+	// Deprecated: use Namespaces instead.
+	Namespace  string   `description:"Sets the namespace used to discover the configuration (Consul Enterprise only)." json:"namespace,omitempty" toml:"namespace,omitempty" yaml:"namespace,omitempty"`
+	Namespaces []string `description:"Sets the namespaces used to discover the configuration (Consul Enterprise only)." json:"namespaces,omitempty" toml:"namespaces,omitempty" yaml:"namespaces,omitempty"`
 }
 
 // SetDefaults sets the default values.
-func (p *Provider) SetDefaults() {
+func (p *ProviderBuilder) SetDefaults() {
 	p.Provider.SetDefaults()
 	p.Endpoints = []string{"127.0.0.1:8500"}
 }
 
+// BuildProviders builds Consul provider instances for the given namespaces configuration.
+func (p *ProviderBuilder) BuildProviders() []*Provider {
+	if p.Namespace != "" {
```

Comment:
```suggestion
    // we can warn about that, because we've already made sure before that
    // Namespace and Namespaces are mutually exclusive
	if p.Namespace != "" {
```